internal/zinc-apiinfo/src/main/scala/xsbt/api/SameAPI.scala

```diff
@@ -258,14 +258,20 @@ class SameAPI(includePrivate: Boolean, includeParamNames: Boolean) {
   def sameType(a: Type, b: Type): Boolean =
     samePending(a, b)(sameTypeDirect)
   def sameTypeDirect(a: Type, b: Type): Boolean =
-    (a, b) match {
-      case (sa: SimpleType, sb: SimpleType)   => debug(sameSimpleTypeDirect(sa, sb), "Different simple types: " + DefaultShowAPI(sa) + " and " + DefaultShowAPI(sb))
-      case (ca: Constant, cb: Constant)       => debug(sameConstantType(ca, cb), "Different constant types: " + DefaultShowAPI(ca) + " and " + DefaultShowAPI(cb))
-      case (aa: Annotated, ab: Annotated)     => debug(sameAnnotatedType(aa, ab), "Different annotated types")
-      case (sa: Structure, sb: Structure)     => debug(sameStructureDirect(sa, sb), "Different structure type")
-      case (ea: Existential, eb: Existential) => debug(sameExistentialType(ea, eb), "Different existential type")
-      case (pa: Polymorphic, pb: Polymorphic) => debug(samePolymorphicType(pa, pb), "Different polymorphic type")
-      case _                                  => differentCategory("type", a, b)
+    {
+      (a, b) match {
+        case (pa: Projection, pb: Projection)       => debug(sameProjection(pa, pb), "Different projection")
+        case (pa: ParameterRef, pb: ParameterRef)   => debug(sameParameterRef(pa, pb), "Different parameter ref")
+        case (sa: Singleton, sb: Singleton)         => debug(sameSingleton(sa, sb), "Different singleton")
+        case (_: EmptyType, _: EmptyType)           => true
+        case (pa: Parameterized, pb: Parameterized) => debug(sameParameterized(pa, pb), "Different parameterized")
+        case (ca: Constant, cb: Constant)           => debug(sameConstantType(ca, cb), "Different constant types: " + DefaultShowAPI(ca) + " and " + DefaultShowAPI(cb))
+        case (aa: Annotated, ab: Annotated)         => debug(sameAnnotatedType(aa, ab), "Different annotated types")
+        case (sa: Structure, sb: Structure)         => debug(sameStructureDirect(sa, sb), "Different structure type")
+        case (ea: Existential, eb: Existential)     => debug(sameExistentialType(ea, eb), "Different existential type")
+        case (pa: Polymorphic, pb: Polymorphic)     => debug(samePolymorphicType(pa, pb), "Different polymorphic type")
```

Comment:
oh the opportunities to backdoor-condition what style of Scala people write, by micro-optimising these parts of the incremental compiler. I'm sure the same opportunities arise in the compiler itself actually.